### DISCUSSION TITLE
Files received

### DISCUSSION THREAD

**Discussion by tobias-carlbom** [jchristn/WatsonTcp#234] [Q&A]
Hi,

I have been playing around with receiving files using file stream. Files are successfully received (in my case from server to client), but i need to verify/make sure that all files actually was received.

In the server code i add metadata for how many files which are to be sent to the client.

In the client code => ServerStreamReceived i tried to append each filename to a list and at the end of the ServerStreamReceived i loop through that list and check if the length are the same as the number of files in the metadata.
Unfortunatly this solution does not seem to work 100%

I also tried a different solution:
In the client code => at the end of ServerStreamReceived i do a sendandwait and send the filename and number of files as metadata to the server.
At the servers SyncResponse SyncRequestReceived i add that file to a list and evaluate the length of the list with the number of files.
This too does not seem to work 100%

Have you guys any ideas for me?

**Comment by jchristn** [comment]
Hi @manprinsen could you provide sample code reduced to highlight the problem?  Cheers

**Comment by tobias-carlbom** [comment]
I is quite a lot of code to share :( So i think i will create a minimalistic project and try to replicate the issue.

But some questions:
Does the code below run asynchronously?
```
                using (FileStream fs = new FileStream(filePath, FileMode.Open, FileAccess.Read))
                {
                    md.Add("filename", Path.GetFileName(filePath));
                    md.Add("md5", Md5File(filePath));
                    md.Add("relativepath", relativePath);
                    md.Add("command", command);
                    md.Add("fileIndex", fileIndex);
                    md.Add("fileCount", fileCount);
                    server.Send(ipPort, len, fs, md);
                }
```

**Reply by jchristn** [reply]
That code runs synchronously.  There are ```SendAsync``` variants that you can use.
